crates/ibc/src/core/ics24_host/identifier.rs

```diff
@@ -123,46 +101,89 @@ impl ChainId {
     /// assert_eq!(ChainId::is_epoch_format("chainA-1"), true);
     /// assert_eq!(ChainId::is_epoch_format("c-1"), true);
+    /// assert_eq!(ChainId::is_epoch_format("-1"), false);
     /// ```
     pub fn is_epoch_format(chain_id: &str) -> bool {
-        let re = safe_regex::regex!(br".*[^-]-[1-9][0-9]*");
-        re.is_match(chain_id.as_bytes())
+        Self::split_chain_id(chain_id).1.is_some()
     }
 
     /// with_version() checks if a chain_id is in the format required for parsing epochs, and if so
     /// replaces it's version with the specified version
     /// ```
     /// use ibc::core::ics24_host::identifier::ChainId;
-    /// assert_eq!(ChainId::new("chainA".to_string(), 1).with_version(2), ChainId::new("chainA".to_string(), 2));
+    /// assert_eq!(ChainId::new("chainA", 1).with_version(2), ChainId::new("chainA", 2));
     /// assert_eq!("chain1".parse::<ChainId>().unwrap().with_version(2), "chain1".parse::<ChainId>().unwrap());
     /// ```
     pub fn with_version(mut self, version: u64) -> Self {
-        if Self::is_epoch_format(&self.id) {
-            self.id = {
-                let mut split: Vec<&str> = self.id.split('-').collect();
-                let version = version.to_string();
-                if let Some(last_elem) = split.last_mut() {
-                    *last_elem = &version;
-                }
-                split.join("-")
-            };
-            self.version = version;
+        if self.version != 0 {
+            if let (name, Some(_)) = Self::split_chain_id(&self.id) {
+                self.id = format!("{name}-{version}");
+                self.version = version;
+            }
         }
         self
     }
+
+    /// Splits chain_id into name and version if the id includes epoch number.
+    ///
+    /// Chain id with epoch number has format `{name}-{version}` where version
+    /// is a positive integer starting with non-zero digit.  If `chain_id` is in
+    /// that format, returns `(name, Some(version))`; otherwise returns
+    /// `(chain_id, None)`.
+    ///
+    /// **Panics** if chain id is in epoch format but the version overflows
+    /// `u64`.
+    fn split_chain_id(chain_id: &str) -> (&str, Option<u64>) {
+        fn split(chain_id: &str) -> Option<(&str, u64)> {
+            let (name, version) = chain_id.rsplit_once('-')?;
+            if name.is_empty() {
+                return None;
+            }
+            let (car, cdr) = version.as_bytes().split_first()?;
+            if b'1' <= *car && *car <= b'9' && cdr.iter().all(u8::is_ascii_digit) {
+                Some((name, version.parse().unwrap()))
+            } else {
+                None
+            }
+        }
+        match split(chain_id) {
+            Some((name, version)) => (name, Some(version)),
+            None => (chain_id, None)
+        }
+    }
+}
+
+impl Default for ChainId {
+    fn default() -> Self {
+        Self { id: String::from(DEFAULT_CHAIN_ID), version: 0 }
```

Comment:
This is invalid id, should be `Self { id: format!("{}-{}", DEFAULT_CHAIN_ID, 0), version: 0 }`

Reply:
It is what the old code was doing:

```rust
    pub fn from_string(id: &str) -> Self {
        let version = if Self::is_epoch_format(id) {
            Self::chain_version(id)
        } else {
            0
        };

        Self {
            id: id.to_string(),
//              ^^ -- id is taken as given
            version,
        }
    }

// ...

impl Default for ChainId {
    fn default() -> Self {
        Self::from_string(DEFAULT_CHAIN_ID)
    }
}
```

Reply:
I know,but is not correct Init. Maybe old code have problem.

Reply:
I’d argue that such change calls for a separate PR especially since it causes test failures.

Reply:
enn，agree to you. Here I think use chainid new method to implement default function.

Reply:
Like I’ve mentioned it breaks existing tests and I won’t have time to work on fixing them.  If you’re not happy with how this PR looks now (and mind you it provides improvements and doesn’t introduce this bug) you’ll have to wait till I find more time to work on this.

Reply:
I don't think we need to worry about the `Default` impl too much, since it is only used in tests. We are also planning on phasing out the use of defaults in tests.